### DISCUSSION TITLE
Memoize value using cache

### DISCUSSION THREAD

**Discussion by nandi95** [laravel/framework#48086] [Ideas]
I was slightly surprised this wasn't included in the framework. Perhaps this would be a worthy addition?
I could send a PR if so.
```php
Cache::macro('memoize', function (string $key, mixed $value): mixed {
    app()->terminating(fn () => Cache::forget($key));

    if (Cache::has($key)) {
        return Cache::get($key);
    }

    $value = value($value);
    Cache::forever($key, $value);

    return $value;
});

**Comment by henzeb** [comment]
You could use the array driver for that.

**Reply by nandi95** [reply]
Array or Memcached whatever, it's more about having this built in to the framework, and the drivers can implement it.
It's another question if this macro should set a driver by default I think.

**Reply by henzeb** [reply]
I think you should install spatie/once or simply do:
```php
Cache::driver('array')->put('key', 'value');
```
The array driver **is** the memoization feature. Writing logic to clear data after terminating just so every driver has a 'memoization' function is overkill and error prone.

**Reply by nandi95** [reply]
@henzeb 
what makes it error prone?

I get your point. The point of the feature is more to make the code more terse:
```php
$admin = Cache::driver('array')->get('admin');
if (!$admin) {
    $admin = User::where('is_admin', true)->firstOrFail();
    Cache::driver('array')->forever('admin', $admin);
}
// vs
$admin = Cache::memoize('admin', static fn () => User::where('is_admin', true)->firstOrFail());
```
perhaps the above can be written like:
```php
Cache::macro('memoize', function (string $key, mixed $value): mixed {
    $cache = Cache::driver('array');
    
    app()->terminating(fn () => $cache->forget($key));
    
    if ($cache->has($key)) {
        return $cache->get($key);
    }
    
    $value = value($value);
    $cache->forever($key, $value);
    
    return $value;
});
```
And in this case if you're right the terminating can even be removed.
I think that  works okay with octane but is the app also torn down in unit testing?
